apps/designer/app/shared/router-utils/canvas-request-params.ts

```diff
@@ -0,0 +1,74 @@
+import config from "~/config";
+
+export type CanvasRouteMode = "edit" | "preview" | "published";
+
+const modes = ["edit", "preview", "published"] as CanvasRouteMode[];
+
+export const getCanvasRequestParams = (request: {
+  url: string;
+  headers: { get: (name: string) => string | null };
+}):
+  | {
+      projectId: string;
+      mode: CanvasRouteMode;
+      pathname: string;
+    }
+  | {
+      projectDomain: string;
+      mode: CanvasRouteMode;
+      pathname: string;
+    }
+  | undefined => {
+  const url = new URL(request.url);
+
+  const pathname = url.pathname;
+
+  if (Object.values(config).some((path) => pathname.startsWith(path))) {
+    return undefined;
+  }
+
+  let projectDomain = undefined;
+  let projectId = undefined;
+
+  const projectIdParam = url.searchParams.get("projectId");
+  if (projectIdParam !== null) {
+    projectId = projectIdParam;
+  }
```

Comment:
nit: here is a one-liner to do the same thing you did in 5 lines
`const projectId =  url.searchParams.get("projectId") ?? undefined`
but also why do you care about null if you never return a null from your function, so you literally won't expose null from the interface

Reply:
Fixed